<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/11-naibolsee-proizvedenie-v-tablice' style="text-decoration: none; color: #1a73e8;">
            11-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 11

В таблице размером `20 × 20` четыре числа, расположенные на одной диагонали, выделены красным.

<img src="../Sources/Project_Euler_011.png" width="500">

Их произведение равно

$$
26 \cdot 63 \cdot 78 \cdot 14 = 1\,788\,696
$$

Каково наибольшее произведение четырех подряд идущих чисел в таблице 20×20, расположенных в любом направлении (вверх, вниз, вправо, влево или по диагонали)?

Ответ: 70600674

## Используемые библиотеки и исходные данные

В данном разборе используется стандартная библиотека Python.

- `perf_counter` — измерение времени выполнения алгоритма.

Переменная `d` содержит исходную таблицу размером `20 × 20`, в которой необходимо найти максимальное произведение четырёх последовательных чисел.

In [4]:
from time import perf_counter

d = [
    [ 8,  2, 22, 97, 38, 15,  0, 40,  0, 75,  4,  5,  7, 78, 52, 12, 50, 77, 91,  8],
    [49, 49, 99, 40, 17, 81, 18, 57, 60, 87, 17, 40, 98, 43, 69, 48,  4, 56, 62,  0],
    [81, 49, 31, 73, 55, 79, 14, 29, 93, 71, 40, 67, 53, 88, 30,  3, 49, 13, 36, 65],
    [52, 70, 95, 23,  4, 60, 11, 42, 69, 24, 68, 56,  1, 32, 56, 71, 37,  2, 36, 91],
    [22, 31, 16, 71, 51, 67, 63, 89, 41, 92, 36, 54, 22, 40, 40, 28, 66, 33, 13, 80],
    [24, 47, 32, 60, 99,  3, 45,  2, 44, 75, 33, 53, 78, 36, 84, 20, 35, 17, 12, 50],
    [32, 98, 81, 28, 64, 23, 67, 10, 26, 38, 40, 67, 59, 54, 70, 66, 18, 38, 64, 70],
    [67, 26, 20, 68,  2, 62, 12, 20, 95, 63, 94, 39, 63,  8, 40, 91, 66, 49, 94, 21],
    [24, 55, 58,  5, 66, 73, 99, 26, 97, 17, 78, 78, 96, 83, 14, 88, 34, 89, 63, 72],
    [21, 36, 23,  9, 75,  0, 76, 44, 20, 45, 35, 14,  0, 61, 33, 97, 34, 31, 33, 95],
    [78, 17, 53, 28, 22, 75, 31, 67, 15, 94,  3, 80,  4, 62, 16, 14,  9, 53, 56, 92],
    [16, 39,  5, 42, 96, 35, 31, 47, 55, 58, 88, 24,  0, 17, 54, 24, 36, 29, 85, 57],
    [86, 56,  0, 48, 35, 71, 89,  7,  5, 44, 44, 37, 44, 60, 21, 58, 51, 54, 17, 58],
    [19, 80, 81, 68,  5, 94, 47, 69, 28, 73, 92, 13, 86, 52, 17, 77,  4, 89, 55, 40],
    [ 4, 52,  8, 83, 97, 35, 99, 16,  7, 97, 57, 32, 16, 26, 26, 79, 33, 27, 98, 66],
    [88, 36, 68, 87, 57, 62, 20, 72,  3, 46, 33, 67, 46, 55, 12, 32, 63, 93, 53, 69],
    [ 4, 42, 16, 73, 38, 25, 39, 11, 24, 94, 72, 18,  8, 46, 29, 32, 40, 62, 76, 36],
    [20, 69, 36, 41, 72, 30, 23, 88, 34, 62, 99, 69, 82, 67, 59, 85, 74,  4, 36, 16],
    [20, 73, 35, 29, 78, 31, 90,  1, 74, 31, 49, 71, 48, 86, 81, 16, 23, 57,  5, 54],
    [ 1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52,  1, 89, 19, 67, 48]
]

## 1) Перебор всех направлений

В данном решении последовательно рассматривается каждая ячейка таблицы как возможное начало последовательности из четырёх чисел.

Для каждой позиции проверяются четыре направления:

- вправо;
- вниз;
- по диагонали вправо вниз;
- по диагонали влево вниз.

Проверять противоположные направления отдельно не требуется, поскольку те же последовательности уже будут рассмотрены с другой начальной позиции.

### Алгоритм

1. Определяется количество строк `r` и столбцов `c`.
2. Последовательно перебираются все элементы таблицы.
3. Если справа достаточно элементов, вычисляется произведение четырёх чисел по горизонтали.
4. Если снизу и справа достаточно элементов, проверяется диагональ вправо вниз.
5. Если снизу достаточно элементов, вычисляется произведение по вертикали.
6. Если снизу и слева достаточно элементов, проверяется диагональ влево вниз.
7. Каждое найденное произведение сравнивается с текущим максимумом.
8. После завершения перебора возвращается максимальное значение.

### Проверка границ

Перед вычислением произведения проверяется, помещается ли последовательность внутри таблицы.

Для движения вправо используется условие:

```python
j < c - 3
```

Для движения вниз:

```python
i < r - 3
```

Для диагонали влево вниз дополнительно проверяется:

```python
j > 2
```

Эти условия предотвращают выход индексов за границы таблицы.

### Особенности

- Каждая ячейка рассматривается как возможная начальная точка.
- Проверяются все необходимые направления.
- Противоположные направления отдельно не обрабатываются.
- Дополнительные структуры данных не используются.
- Временная сложность составляет `O(r · c)`.
- Пространственная сложность составляет `O(1)`.

### Вывод

Последовательный обход таблицы позволяет проверить все возможные расположения четырёх соседних чисел. Благодаря проверке только четырёх направлений и контролю границ решение остаётся компактным и эффективным.

In [9]:
def get_result_1(nums: list[list[int]]) -> int:
    m = 0
    r, c = len(nums), len(nums[0])
    for i in range(r):
        for j in range(c):
            if j < c - 3:
                m = max(m, nums[i][j] * nums[i][j + 1] * nums[i][j + 2] * nums[i][j + 3])
                if i < r - 3:
                    m = max(m, nums[i][j] * nums[i + 1][j + 1] * nums[i + 2][j + 2] * nums[i + 3][j + 3])

            if i < r - 3:
                m = max(m, nums[i][j] * nums[i + 1][j] * nums[i + 2][j] * nums[i + 3][j])
                if j > 2:
                    m = max(m, nums[i][j] * nums[i + 1][j - 1] * nums[i + 2][j - 2] * nums[i + 3][j - 3])                      
    return m



start = perf_counter()
print(get_result_1(d))
end = perf_counter()

print(f'{end - start:.5f}')

70600674
0.00061


## 2) Использование направлений

В данном решении используется тот же алгоритм, однако вместо отдельных проверок для каждого направления применяется единый список направлений.

Каждое направление задаётся парой смещений `(dr, dc)`, после чего произведение четырёх чисел вычисляется одинаковым способом независимо от ориентации.

### Отличия от предыдущего решения

- Все направления объединены в одну коллекцию.
- Проверка границ выполняется по единому правилу.
- Для вычисления произведения используется функция `prod()`.
- Код становится более компактным и легче расширяется.
- Алгоритм и временная сложность остаются без изменений.

### Вывод

Использование списка направлений позволяет избавиться от повторяющегося кода и сделать решение более универсальным.

In [10]:
from math import prod


def get_result_2(nums: list[list[int]]) -> int:
    rows = len(nums)
    cols = len(nums[0])

    directions = (
        (0, 1),   # вправо
        (1, 0),   # вниз
        (1, 1),   # диагональ вправо вниз
        (1, -1),  # диагональ влево вниз
    )

    mx = 0
    for row in range(rows):
        for col in range(cols):
            for dr, dc in directions:
                if 0 <= row + 3 * dr < rows and 0 <= col + 3 * dc < cols:
                    mx = max(mx, prod(nums[row + step * dr][col + step * dc] for step in range(4)))

    return mx


start = perf_counter()
print(get_result_2(d))
end = perf_counter()

print(f'{end - start:.5f}')

70600674
0.00161


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **1000 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [12]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(d), number=1000) / 1000
    print(f'Алгоритм {i}: {tm:.7f}')

Алгоритм 1: 0.0003179
Алгоритм 2: 0.0012966


## Заключение

В задаче были рассмотрены два подхода к поиску максимального произведения четырёх последовательных чисел в таблице.

Первый алгоритм проверяет каждое направление отдельным условием и напрямую перемножает четыре нужных элемента. Такой код содержит повторяющиеся выражения, но выполняет минимальное количество дополнительных операций.

Во втором решении направления объединены в одну коллекцию, а произведение вычисляется универсально с помощью функции `prod()` и генераторного выражения. Благодаря этому код становится компактнее, понятнее и легче расширяется, однако появляются дополнительные накладные расходы.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `0.0003179` с
- **Алгоритм 2:** `0.0012966` с

Второй алгоритм оказался примерно в **4 раза медленнее** первого по нескольким причинам:

- для каждой ячейки дополнительно перебирается коллекция направлений;
- многократно вычисляются координаты `row + step * dr` и `col + step * dc`;
- для каждого подходящего направления создаётся генераторное выражение;
- вызывается универсальная функция `prod()`;
- выполняется внутренний цикл `range(4)` вместо четырёх прямых обращений к элементам таблицы.

Оба решения имеют одинаковую временную сложность `O(r · c)`, поскольку число направлений и длина последовательности являются константами. Разница возникает не из-за асимптотики, а из-за стоимости дополнительных операций внутри каждого шага.

Таким образом, первое решение лучше по скорости, а второе — по универсальности и удобству сопровождения. Для фиксированной задачи с четырьмя заранее известными направлениями прямые вычисления оказываются эффективнее обобщённой реализации.